protocols/bftcosi/bftcosi.go

```diff
@@ -393,29 +379,112 @@ func (bft *ProtocolBFTCoSi) handleChallengeCommit(ch *ChallengeCommit) error {
 
 	// store the exceptions for later usage
 	bft.tempExceptions = ch.Signature.Exceptions
-	log.Lvl4("BFTCoSi handle Challenge COMMIT")
 
 	if bft.IsLeaf() {
 		return bft.handleResponseCommit(nil)
 	}
 
-	if err := bft.SendToChildrenInParallel(ch); err != nil {
-		log.Error(err)
+	return bft.SendToChildrenInParallel(&ch)
+}
+
+// handleResponse is called when a response message arrives.
+func (bft *ProtocolBFTCoSi) handleResponse(msgs []responseChan) error {
+	for _, msg := range msgs {
+		log.Lvl3(bft.Name(), "RoundType:", msg.Response.TYPE)
+		if bft.closing {
+			return errors.New("Quitting instance")
+		}
+		var resp *Response
+		if !bft.IsLeaf() {
+			resp = &msg.Response
+		}
+		if msg.Response.TYPE == RoundPrepare {
+			if err := bft.handleResponsePrepare(resp); err != nil {
+				return err
+			}
+		} else {
+			if err := bft.handleResponseCommit(resp); err != nil {
+				return err
+			}
+		}
 	}
 	return nil
 }
 
+// startAnnouncementPrepare create its announcement for the prepare round and
+// sends it down the tree.
+func (bft *ProtocolBFTCoSi) startAnnouncement(t RoundType) error {
+	log.Lvl3(bft.Name(), "RoundType:", t)
+	bft.announceChan <- announceChan{Announce: Announce{TYPE: t}}
+	return nil
+}
+
+// startCommitment sends the first commitment to the parent node
+func (bft *ProtocolBFTCoSi) startCommitment(t RoundType) error {
+	log.Lvl3(bft.Name(), "RoundType:", t)
+	cm := bft.getCosi(t).CreateCommitment(nil)
+	return bft.SendToParent(&Commitment{TYPE: t, Commitment: cm})
+}
+
+// startChallenge creates the challenge and sends it to its children
+func (bft *ProtocolBFTCoSi) startChallenge(t RoundType) error {
+	log.Lvl3(bft.Name(), "RoundType:", t)
+	switch t {
+	case RoundPrepare:
+		// need to hash the message before so challenge in both phases are not
+		// the same
+		data := sha512.Sum512(bft.Msg)
+		ch, err := bft.prepare.CreateChallenge(data[:])
+		if err != nil {
+			return err
+		}
+		bftChal := &ChallengePrepare{
+			Challenge: ch,
+			Msg:       bft.Msg,
+			Data:      bft.Data,
+		}
+
+		bft.challengePrepareChan <- challengePrepareChan{ChallengePrepare: *bftChal}
+	case RoundCommit:
+		// commit phase
+		ch, err := bft.commit.CreateChallenge(bft.Msg)
+		if err != nil {
+			return err
+		}
+
+		// send challenge + signature
+		cc := &ChallengeCommit{
+			Challenge: ch,
+			Signature: &BFTSignature{
+				Msg:        bft.Msg,
+				Sig:        bft.prepareSignature,
+				Exceptions: bft.tempExceptions,
+			},
+		}
+		bft.challengeCommitChan <- challengeCommitChan{ChallengeCommit: *cc}
+	}
+	return nil
+}
+
+// startResponse dispatches the response to the correct round-type
+func (bft *ProtocolBFTCoSi) startResponse(t RoundType) error {
+	log.Lvl3(bft.Name(), "RoundType:", t)
+	bft.handleResponse([]responseChan{{Response: Response{TYPE: t}}})
```

Comment:
Not sure if this would work with https://github.com/dedis/cothority/pull/511/files#r74425842